lib/signon/client.rb

```diff
@@ -4,74 +4,90 @@
-        res = do_request(req, uri)
-        raise TokenNotCreated, "Status: #{res.code}; #{res.message}; #{res.body}" unless %w[200 201].include?(res.code)
-
-        JSON.parse(res.body).fetch("token")
+        req = build_post_request(
+          "/api-users/#{api_user_id}/authorisations",
+          {
+            application_id: application_id,
+            permissions: permissions,
+          },
+        )
+        res = do_request(req)
+        raise TokenNotCreated, "Status: #{res.code}; #{res.message}; #{res.body}" unless req_successful(res)
+
+        JSON.parse(res.body)
       end
     end
 
-    def test_bearer_token(api_user:, token:, application_name:, permissions:)
+    def test_bearer_token(api_user_id:, application_id:, token:, permissions:)
       attempt do
-        uri = URI("#{@api_url}/authorisations/test")
-        req = Net::HTTP::Post.new(uri)
-        req.body = {
-          token: token,
-          api_user_email: api_user,
-          application_name: application_name,
-          permissions: permissions,
-        }.to_json
-        res = do_request(req, uri)
+        req = build_post_request(
+          "/api-users/#{api_user_id}/authorisations/test",
+          {
+            token: token,
+            application_name: application_id,
+            permissions: permissions,
+          },
+        )
+        res = do_request(req)
         raise TokenNotFound, "Status: #{res.code}; #{res.message}; #{res.body}" unless res.code == "200"
       end
     end
 
     def create_application(name:, description:, home_uri:, permissions:, redirect_uri:)
-      uri = URI("#{@api_url}/applications")
-      req = Net::HTTP::Post.new(uri)
-      req.body = {
+      req = build_post_request("/applications", {
         name: name,
         description: description,
         home_uri: home_uri,
         permissions: permissions,
         redirect_uri: redirect_uri,
-      }.to_json
-      res = do_request(req, uri)
-      raise ApplicationAlreadyCreated if already_exists(res)
-      raise ApplicationNotCreated, "Status: #{res.code}; #{res.message}; #{res.body}" unless %w[200 201].include?(res.code)
+      })
+      res = do_request(req)
+      raise ApplicationAlreadyCreated if res.code_type == Net::HTTPConflict
+      raise ApplicationNotCreated, "Status: #{res.code}; #{res.message}; #{res.body}" unless req_successful(res)
 
       JSON.parse(res.body)
     end
 
     def get_application(name:)
       uri = URI("#{@api_url}/applications?name=#{CGI.escape name}")
       req = Net::HTTP::Get.new(uri)
-      res = do_request(req, uri)
+      res = do_request(req)
       raise ApplicationNotFound, "Status: #{res.code}; #{res.message}; #{res.body}" unless res.code == "200"
 
       JSON.parse(res.body)
     end
 
+    def create_api_user(name:, email:)
+      req = build_post_request("/api-users", { name: name, email: email })
+      res = do_request(req)
+      raise ApiUserAlreadyCreated if res.code_type == Net::HTTPBadRequest
```

Comment:
Was it intentional that the API returns 400 when trying to create an already-existing ApiUser but 409 on trying to create an already-existing Application? Is it a legacy thing that's a pain to fix?

Reply:
Yes, thanks for flagging that. I've added a commit to the Signon API PR to fix this annoying inconsistency and added some detail in the commit message (https://github.com/alphagov/signon/pull/1768/commits/eb71ff82b173f615eee06e6028916e0be9ef3879).

Essentially the ApiUser and the Application models are intertwined with different gems (Devise and Doorkeeper respectively) which validate models inconsistently, which has leaked into this status code. The commit I've put up should resolve that inconsistency and I can now change the client to check for a 409.

Reply:
Cool, thanks for fixing that! 💛